### ISSUE TITLE
regex in loop triggers on OnceLock::<Regex>::get_or_init in loop

### ISSUE THREAD

**Issue by hkBst** [rust-lang/rust-clippy#15237]
### Summary

Due to the nature of OnceLock, initialization happens only once, so I think this is a false positive:

### Lint Name

regex_creation_in_loops

### Reproducer

I tried this code:

```rust
macro_rules! static_regex {
    ($re:literal) => {{
        static RE: ::std::sync::OnceLock<::regex::Regex> = ::std::sync::OnceLock::new();
        RE.get_or_init(|| ::regex::Regex::new($re).unwrap())
    }};
}

fn main() {
    while false {
        if static_regex!(r"^\s*//").is_match("") {
            break;
        }
    }
}
```

I saw this happen:
```
warning: compiling a regex in a loop
  --> src/main.rs:4:27
   |
4  |         RE.get_or_init(|| ::regex::Regex::new($re).unwrap())
   |                           ^^^^^^^^^^^^^^^^^^^
...
10 |         if static_regex!(r"^\s*//").is_match("") {
   |            ------------------------ in this macro invocation
   |
help: move the regex construction outside this loop
  --> src/main.rs:9:5
   |
9  |     while false {
   |     ^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#regex_creation_in_loops
   = note: `#[warn(clippy::regex_creation_in_loops)]` on by default
   = note: this warning originates in the macro `static_regex` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see this happen:
no warning

### Version

current nightly

**Comment by samueltardieu** [comment]
You'll still check in a loop whether the regex has been built or not, while you could have gotten the regex inside a variable before the loop and use it inside the loop.

**Comment by hkBst** [comment]
Yes, I suppose you could argue that using `get_or_init` in a loop is also pretty silly, but it doesn't have anything to do with regexes...
